workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/WorkflowRendering.kt

```diff
@@ -0,0 +1,201 @@
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalLifecycleOwner
+import androidx.compose.ui.viewinterop.AndroidView
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.State.DESTROYED
+import androidx.lifecycle.Lifecycle.State.INITIALIZED
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+import androidx.lifecycle.ViewTreeLifecycleOwner
+import com.squareup.workflow1.ui.Compatible
+import com.squareup.workflow1.ui.ViewEnvironment
+import com.squareup.workflow1.ui.ViewFactory
+import com.squareup.workflow1.ui.ViewRegistry
+import com.squareup.workflow1.ui.WorkflowLifecycleOwner
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.WorkflowViewStub
+import com.squareup.workflow1.ui.getFactoryForRendering
+import com.squareup.workflow1.ui.getShowRendering
+import com.squareup.workflow1.ui.showRendering
+
+/**
+ * Renders [rendering] into the composition using this [ViewEnvironment]'s [ViewRegistry] to
+ * generate the view.
+ *
+ * This function fulfills a similar role as [WorkflowViewStub], but is much more convenient to use
+ * from Composable functions.
+ *
+ * ## Example
+ *
+ * ```
+ * data class FramedRendering(
+ *   val borderColor: Color,
+ *   val child: Any
+ * )
+ *
+ * val FramedContainerViewFactory = composeViewFactory<FramedRendering> { rendering, environment ->
+ *   Surface(border = Border(rendering.borderColor, 8.dp)) {
+ *     WorkflowRendering(rendering.child, environment)
+ *   }
+ * }
+ * ```
+ *
+ * @param rendering The workflow rendering to display. May be of any type for which a [ViewFactory]
+ * has been registered in this environment's [ViewRegistry].
+ * @param modifier A [Modifier] that will be applied to composable used to show [rendering].
+ *
+ * @throws IllegalArgumentException if no factory can be found for [rendering]'s type.
+ */
+@WorkflowUiExperimentalApi
+@Composable public fun WorkflowRendering(
+  rendering: Any,
+  viewEnvironment: ViewEnvironment,
+  modifier: Modifier = Modifier
+) {
+  val viewRegistry = remember(viewEnvironment) { viewEnvironment[ViewRegistry] }
+
+  // This will fetch a new view factory any time the new rendering is incompatible with the previous
+  // one, as determined by Compatible. This corresponds to WorkflowViewStub's canShowRendering
+  // check.
+  // We also use the ViewRegistry as a key, in case a new one gets passed down and we need to ask
+  // it for a more current factory.
+  // The view registry may return a new factory instance for a rendering every time we ask it, for
+  // example if an AndroidViewRendering doesn't share its factory between rendering instances. We
+  // intentionally don't ask it for a new instance every time to match the behavior of
+  // WorkflowViewStub and other containers, which only ask for a new factory when the rendering is
+  // incompatible.
```

Comment:
WVS seems to think the contract of `ViewRegistry` is that it will always return the same factory for the total set of all renderings that are compatible with each other, and only asks for a new factory if that changes. That is:
```kotlin
if (view.canShowRendering(newRendering)) {
  view.showRendering(newRendering)
  return
}

// Else, ask for a new ViewFactory and initialize it.
```

I think changing this semantic could be significant, and I'd like to not do that here and instead stay consistent. One of the nice properties of this is that view factories' identify doesn't matter. E.g. the proposed `asComposeViewFactory` would return a different instance every time it's called – with the current semantic, we'd only call it once, and we would only ever care about or use the one returned from that call. With a more aggressive check, we'd be creating a new wrapper on every update, and then that wrapper would have to implement equality itself.

I think the current semantic is simpler and easier to reason about.